design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,147 @@
+will will also perform whatever Ceph operations are necessary to remove the OSD as would be expected
+by the user. This will replace the manual, imperative steps currently required by Rook-Ceph users to
+remove OSDs from the storage cluster documented
+[here for Rook v1.2](https://rook.io/docs/rook/v1.2/ceph-osd-mgmt.html#remove-an-osd).
+
+There is currently a proposal for a shared Ceph Mgr Orchestrator Module implementing common OSD
+removal steps [[link]](https://github.com/ceph/ceph/pull/32677) usable by any other Ceph Mgr
+Orchestrator module. The OSD removal mechanism should use this where possible.
+
+Rook will not automatically remove Kubernetes Deployments for OSDs unless
+`removeOSDsIfOutAndSafeToRemove: true`. This should be documented for users, and users should be
+advised to set this value to `true`.
+
+The `CephCluster` CRD allows users to add disks to nodes in several places including disks on all
+nodes in the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different
+ways). To avoid the complexity of having to manage the logic for determining how to remove a node
+from the\ `CephCluster` resource, OSD removal will **not** alter the `CephCluster` resource. While
+doing so might be convenient for users, there are a lot of opportunities in development to miss
+corner cases or introduce logic bugs.
+
+Since the Rook Orchestrator Mgr Module will not remove user-configured items from the `CephCluster`
+resource's storage spec, if the `CephCluster` resource is configured to add OSDs by the user **and**
+a removed disk is zapped, Rook could re-add the disk after the disk is removed via the Ceph
+Orchestrator interface. To prevent user confusion and frustration, OSD removal should **not** zap
+disks so that they are not added back into the Ceph cluster automatically by Rook. The user should
+be made aware that Rook can automatically re-add removed OSDs if they are zapped manually by the
+user via documentation to avoid possible confusion and annoyance.
+
+Users should be advised in documentation that if they plan to use the Ceph CLI or Dashboard to
+manage OSD addition in the Rook-Ceph cluster, it is recommended to set `useAllDevices: false`, leave
+both disk pattern matching empty, and to not specify any other disks in the `CephCluster` resource
+to avoid giving Rook the ability to re-add OSDs when the user does not wish it to.
+
+### Not in scope
+It is **not** within the scope of this design to add or remove nodes from the `CephCluster`
+resource. If a node does not exist in the `CephCluster` resource, the user should not be allowed to
+run OSD provisioning jobs on the node, and if the user removes all OSDs from a node, that node
+should remain in the `CephCluster` resource assuming to have been added by an orchestrator add node
+command.
+
+It is **not** the Rook Orchestrator Mgr Module's job to ensure that disks are added or removed from
```

Comment:
It does seem appropriate that when driveGroups are being generated by the mgr module, the mgr module would overwrite any other settings that might be in the `node` section. It would be more complicated to support any existing settings under `node` elements. Those settings don't seem relevant when you're defining driveGroups. Is that along the same lines as what's you're saying in this paragraph? Or are you saying something different?

Reply:
I am proposing that the mgr module should **not** alter anything in the CRD other than to add Drive Groups (or remove Drive Groups that it has created). It should not delete or alter any user-specified values, and it should not add anything other than Drive Groups.

Reply:
While we could choose to be careful about not overwriting other settings in the CR, I wonder if it just isn't necessary to support it. If the mgr module is generating the CR, I would imagine that the user is not going to be independently be updating the CR. The mgr module could be opinionated and declare that it is taking over the `storage` section of the CR because driveGroups are going to cause conflicts with other storage settings or at least cause confusing overlaps.

Reply:
Good points. I added this to the discussion for Monday. We certainly could make the mgr opinionated, and I am starting to see how that might really just be a better idea.